Intel ADSP: Use features to define conditional code

edersondisouza opened the pull request:
Instead of using SoC version to decide which code to run on "ifdefs", declare features on DTS and use this information instead.
While at it, did some other minor cleanup.

edersondisouza committed:
drivers/dai/intel/ssp: Decide about clock gating based on Kconfig

Instead of using version of the SoC, declare on Kconfig the need for it,
and use this information to decide upon enabling the code or not.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
drivers/dai/intel/ssp: Get number of SSP ports from Kconfig

Instead of using SoC versions, define the information about base and
extended ports on Kconfig, and use this information from there.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
drivers/mm: Get some bit configurations from DTS instead of SoC version

Migrate information to DTS and get it from there on the code. Note that
for CAVS 15, the information is not migrated as there's no DTS entry for
it. It can be brought back (in the DTS) if TLB support is enabled for
it.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
drivers/mm: Remove redundant ifdef for MTL ADSP TBL implementation

An `ifdef CONFIG_SOC_SERIES_INTEL_ACE1X` currently always evaluates to
true, as `drivers/mm/mm_drv_intel_adsp_mtl_tlb.c` is only built when
`CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB=y`, which depends on
`CONFIG_SOC_SERIES_INTEL_ACE1X`. So remove the ifdef.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
soc/xtensa/intel_adsp: Get register address from DTS

Instead of hardcoding it where it's used.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
soc/xtensa/intel_adsp: Disable L2 cache at boot based on Kconfig

Instead of deciding based on SoC version, add information to Kconfig and
get it from there on code.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

galak commented on dts/bindings/sram/intel,adsp-settings.yaml:
```diff
@@ -0,0 +1,22 @@
+
+description: Intel ADSP SRAM settings.
+
+compatible: "intel,sram-settings"
```
Would these not be properties of some SRAM node?

edersondisouza replied:
That makes some sense, at least for the hp-sram and l2 bits, which have nodes. Not sure yet about the L1 side - will look into that.

ceolin replied:
Agree with @galak, it looks to me like a device properties

edersondisouza replied:
For the "init HP SRAM" that was fairly easy, so done on v2. But I kept the others ("disable l2 cache at boot" and "power on SRAM cache") in this "global" way as there's no single node associated with what these config. Further thoughts?

kv2019i reviewed (dismissed):
I'll let other reviewers cover the DTS definitions (whether they are in appropriate places), but changes in driver code look good and overall this cleans up the driver code. Thanks!

ceolin commented on dts/bindings/i2s/intel,ssp-ports.yaml:
```diff
@@ -0,0 +1,18 @@
+
+    extended:
+      type: int
+      description: Number of extended SSP ports available.
```
Wouldn't be better having ssp-clk-gating and ssp-ports merged ? I think that is preety similar to `adsp-settings`.

edersondisouza replied:
I'm not entirely sure `adsp-settings` is ideal, being a catch all for "global settings". Aren't these specific binds clearer?

MaureenHelm replied:
Yeah, these ssp properties all need to be part of one compatible.

ceolin commented:
PR is looking nice, that is a really good direction.

edersondisouza commented:
v2:
 - Move "init HP SRAM" setting to a node;
 - Remove another useless ifdef.

MaureenHelm reviewed (changes requested):
I like the drivers/mm changes, but I think some of the new ssp properties need to be combined into one binding. It doesn't make sense to create a new binding for each new property. A binding corresponds to a hardware component, like an IP block, and properties describe features of that hardware.

edersondisouza commented on dts/bindings/i2s/intel,ssp-clk-gating.yaml:
```diff
@@ -0,0 +1,14 @@
+
+description: Intel SSP clock gating
+
+compatible: "intel,ssp-clk-gating"
```
Ok. The existing one is `intel,ssp-dai`, and this one is already being used by each ssp node (`ssp0` to `ssp5`). I'll give it a try, but that means that this will be duplicated for all nodes (and in the code I'll get them from an arbitrary one, such as `ssp0`). I wonder if we don't/shouldn't have "class" properties for all similar nodes.

edersondisouza replied:
Or are you talking about having an `intel,ssp-settings`, more like @ceolin comment, @MaureenHelm?

MaureenHelm replied:
Possibly, yes. Can you look at syscon as another option?

edersondisouza replied:
`syscon` seems more related to misc registers, not specific properties of a given hardware version, so I don't really think it helps here.

nashif commented on dts/bindings/sram/intel,sram0.yaml:
```diff
@@ -0,0 +1,18 @@
+    reg:
+      required: true
+
+    need-init:
```
This looks more like Kconfig material and probably should not be in DTS.

edersondisouza replied:
@nashif and the others do still look like more DTS? What is about this one that makes it more Kconfig-ish? After all, it's still a "property" from the hardware - I want to better tune my criteria here.

nashif replied:
The usage in the existing code and here is a bit cumbersome, i.e. if you have the sram already configured in the system and enable, we probably do not need this property, if it is enabled, it will be initialized, platform that do not have this, will not initialize it.
Not only this one, but also `    need-power-on-cache`, I commented on by forgot to push.

edersondisouza commented:
v3:
  - A single `ssp-settings` for the "global" ssp properties;
  - Fixed (some) spacing issues.

MaureenHelm commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -83,6 +83,12 @@
 		zephyr,memory-region = "IMR1";
 	};
 
+	ssp_ports: ssp-ports {
+		compatible = "intel,ssp-ports";
+		base = <6>;
+		extended = <0>;
+	};
+
```
Leftover from previous revision? This should be ssp-settings now

edersondisouza replied:
Oops!

edersondisouza commented:
v4:
 - Rebased;
 - Fixed leftover changes.

edersondisouza committed:
soc/xtensa/intel_adsp/common: Power on cache SRAM banks based on Kconfig

Instead of using SoC version to decide to power on or not, add this
information to Kconfig and get it from there.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
soc/xtensa/intel_adsp/cavs: Init HP SRAM based on Kconfig

Instead of deciding to init based on SoC version, define that
information in Kconfig and use it from there.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza commented:
v5:
  - HP SRAM init and SRAM cache power on configs migrated to Kconfig instead, following @nashif  suggestion;

This has a bit less lines, but hidden Kconfig trickery doesn't look exactly natural. Maybe have the DTS option, but use the Kconfig to change into `CONFIG_XXX` to be used from code? Or is this the worst of both worlds? =D

MaureenHelm commented on drivers/dai/intel/ssp/ssp.h:
```diff
@@ -235,6 +235,11 @@
 #define SHIM_CLKCTL_I2SEFDCGB(x)	BIT(18 + x)
 
+/* Is dynamic clock gating a must? */
+#if DT_PROP(DT_NODELABEL(ssp_settings), force_dynamic_clock_gating)
```
We need to stop using hard-coded nodelabels and replace them with phandle references.

MaureenHelm commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -46,11 +46,12 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
  * Number of significant bits in the page index (defines the size of
  * the table)
  */
+#define TLB_PADDR_SIZE DT_PROP(DT_NODELABEL(tlb), paddr_size)
+#define TLB_EXEC_BIT   BIT(DT_PROP(DT_NODELABEL(tlb), exec_bit_idx))
+#define TLB_WRITE_BIT  BIT(DT_PROP(DT_NODELABEL(tlb), write_bit_idx))
```
Ditto about hard-coded nodelabels

MaureenHelm commented on soc/xtensa/intel_adsp/cavs/include/intel_apl_adsp/adsp_memory.h:
```diff
@@ -42,4 +42,13 @@
 
 #define ADSP_L1_CACHE_PREFCTL_VALUE 0
 
+#define ADSP_L2CC_ADDR          0x1500
```
The base address should come from devicetree

edersondisouza committed:
tests/boards/intel_adsp: Fix SSP test build

Test wasn't building since 8dd57467c made Intel DAI SSP depend on power
management. This patch fix this by adding the required configs to the
test prj.conf.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza commented:
v6:
 - Fix Intel ADSP SSP test;
 - Get `ssp-settings` node via phandle on first SSP instance;
 - Get TLB info using DT_DRV_COMPAT instance instead of hardcoding nodelabel;
 - Declare L2CC (L2 Cache Control) address on DTS.

ceolin commented on dts/bindings/i2s/intel,ssp-settings.yaml:
```diff
@@ -0,0 +1,22 @@
+
+    ports-extended:
+      type: int
+      description: Number of extended SSP ports available.
```
It is not a blocker for me, but I really find weird dt using for settings. DT should describe HW and general settings should come from Kconfig IMHO.

edersondisouza replied:
Well, isn't the number of ports a property of the HW? Before this patch, it was being decided on HW version alone...

MaureenHelm commented on drivers/dai/intel/ssp/ssp.c:
```diff
@@ -34,6 +34,11 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 #define SSP_ARRAY_INDEX(dir) dir == DAI_DIR_RX ? DAI_DIR_CAPTURE : DAI_DIR_PLAYBACK
 
+/* Is dynamic clock gating a must? */
+#if DT_INST_PROP_BY_PHANDLE(0, settings, force_dynamic_clock_gating)
+#define SSP_CLK_FORCE_DYNAMIC_CLOCK_GATING 1
+#endif
```
Store this in `struct dai_config`

edersondisouza replied:
I'm not sure I get why... this configuration only defines if some code is going to be compiled or not - why store it in runtime?

MaureenHelm replied:
To avoid hard-coding the instance number

edersondisouza replied:
But the hard-coded instance number is zero, which is kinda safe, no? Any instance could be linked to the information, using the first one seems reasonable - I think I saw that usage elsewhere.
I also wonder if we are trying to satisfy too many constraints here: avoid using DT_NODELABEL directly, avoid hardcoding an instance number. Although I still think those are hardware properties and should be described in DTS, it seems that by moving to Kconfig would allow us workaround those restrictions.

MaureenHelm replied:
Instance numbers are not safe. Hardcoding a nodelabel and hardcoding an instance number are two variations of the same problem.

edersondisouza replied:
Ok, will see if there's any way to salvage DTS here, and if not, I'll move those to Kconfigs.

MaureenHelm commented on drivers/dai/intel/ssp/ssp.h:
```diff
@@ -40,17 +40,10 @@
-#if CONFIG_SOC_INTEL_CAVS_V15
 /** \brief Number of 'base' SSP ports available */
-#define DAI_INTEL_SSP_NUM_BASE			4
+#define DAI_INTEL_SSP_NUM_BASE		DT_INST_PROP_BY_PHANDLE(0, settings, ports_base)
```
Store this in `struct dai_config`

edersondisouza replied:
I'm also not sure I get why... while this configuration is also used in runtime, it's also used to define the size of an array and it seems fairly intel ssp specific - why would it belong to the generic struct?

MaureenHelm replied:
Fair point about being Intel SSP specific. Is there another way to avoid hard-coding the instance number?

edersondisouza commented:
v7:
 - Using mainly Kconfig for the code conditionals;
 - DTS usage reduced to two patches - one register address and some bits of TLB information;
 - Rebased.

(CI issue is unrelated, will retry when CI issue is fixed.)

MaureenHelm commented on drivers/dai/intel/ssp/Kconfig.ssp:
```diff
@@ -17,3 +17,11 @@ config DAI_INTEL_SSP
+
+config DAI_SSP_CLK_FORCE_DYNAMIC_CLOCK_GATING
+	bool
+	default n
```
Remove this line. Boolean configs are implicitly default n.

MaureenHelm reviewed (changes requested):
One nit, but otherwise LGTM

edersondisouza commented:
v8:
 - Removed redundant defaults from Kconfig;
 - Rebased.
(CI issue seems unrelated, will retry when CI issue is fixed.)
